yarn-project/noir-contracts/src/contracts/easy_private_voting_contract/src/main.nr

```diff
@@ -0,0 +1,117 @@
+    tally: Map<PublicState<Field, FIELD_SERIALIZED_LEN>>, // we will store vote as key and number of votes as value
+    voteEnded: PublicState<bool, BOOL_SERIALIZED_LEN>, // voteEnded is boolean
+}
+    // docs:end:storage_struct
+    // docs:start:storage_impl
+    impl Storage {
+	fn init(context: Context) -> Self {
+	    Storage {
+		 admin: PublicState::new(
+		    context, 
+            1, // storage slot. this can be anything except 0. it is hashed, and hash on 0 = 0
+            AztecAddressSerializationMethods,
+		), 
+        tally: Map::new(
+		    context, 
+            2, 
+            |context, slot| {
+			PublicState::new(
+			    context,
+                slot, 
+                FieldSerializationMethods,
+			)
+		    },
+		 ),
+		  voteEnded: PublicState::new(
+		    context, 
+            3, 
+            BoolSerializationMethods,
+		)
+	    } }
+	}
+    // docs:end:storage_impl
+    // docs:start:constructor
+    #[aztec(private)] // annotation to mark function as private and expose private context
+    fn constructor(admin: AztecAddress) { // called when contract is deployed
+        context.call_public_function(
+            // we cannot update public state directly from private function but we can call public function (which queues it)
+            context.this_address(),// address that calls public function. here we are using contract address for privacy
+            compute_selector("_initialize((Field))"), // selector of  set_admin()
+            [admin.address] // parameters for set_admin()
+        );
+    }
+    // docs:end:constructor
+    // docs:start:initialize
+    #[aztec(public)] // annotation to mark function as public and expose public context
+    internal fn _initialize(admin: AztecAddress) { // internal - can only be called by contract
+        storage.admin.write(admin);
+        storage.voteEnded.write(false);
+    }
+    // docs:end:initialize
+    // docs:start:cast_vote
+    #[aztec(private)] // annotation to mark function as public and expose private context
+    fn cast_vote(voteId: Field) {
+     // docs:end:cast_vote
+     // docs:start:nullifier
+        let secret = get_secret_key(context.msg_sender()); // get secret key of caller of function
+        let nullifier = dep::std::hash::pedersen_hash([context.msg_sender(), secret.low, secret.high]); // compute nullifier with this secret key so others can't descrypt it
+        context.push_new_nullifier(nullifier, EMPTY_NULLIFIED_COMMITMENT); // push nullifier
+    // docs:end:nullifier
+    // docs:start:call_add_to_tally_public
+        context.call_public_function(
+            context.this_address(), 
+            compute_selector("add_to_tally_public(Field)"), 
+            [voteId]);
+    }
+    // docs:end:call_add_to_tally_public
+
+    // docs:start:add_to_tally_public
+    #[aztec(public)]
+    internal fn add_to_tally_public(voteId: Field) {
+    // docs:end:add_to_tally_public
+    // docs:start:assert_vote_not_ended
```

Comment:
nit - can we make it so the entire function in one //docs:start thingy just so code doesn't look ugly for internal engineers

Also reduces the odds of messups

And then you can have a para explaining line by line